Added nullptr checks for RenderObject

pratiksha-99 opened the pull request:
#### ac941d96e47907797eb4f0c7fb352164b5512873
<pre>
Added nullptr checks for RenderObject
Need the bug URL (OOPS!).
<a href="https://rdar.apple.com/135596465">rdar://135596465</a>.

Reviewed by NOBODY (OOPS!).

Added nullptr checks for RenderObject before de-referencing.

* LayoutTests/fast/repaint/repaint-renderer-with-layout-crash-expected.txt: Added.
* LayoutTests/fast/repaint/repaint-renderer-with-layout-crash.html: Added.
* Source/WebCore/rendering/RenderObject.cpp:
 (WebCore::RenderObject::propagateRepaintToParentWithOutlineAutoIfNeeded const):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ac941d96e47907797eb4f0c7fb352164b5512873

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/69923 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49324 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/22676 "Built successfully") | [⏳ 🛠 wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/21081 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/72040 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/57124 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/20932 "Built successfully") | [⏳ 🧪 wpe-wk2 ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13970 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/72989 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44937 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60314 "Passed tests") | [⏳ 🧪 api-wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/41602 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/17743 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/19458 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/63527 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18093 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/75723 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/14148 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/17320 "20 flakes 1 failures") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/63189 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/14184 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60389 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63122 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11144 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4753 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/45127 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/46201 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/47472 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/45942 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -981,7 +981,10 @@ void RenderObject::propagateRepaintToParentWithOutlineAutoIfNeeded(const RenderL
+        if (!renderer || !originalRenderer)
+            return;
+
+        bool rendererHasOutlineAutoAncestor = renderer->hasOutlineAutoAncestor() || originalRenderer->hasOutlineAutoAncestor(); // PCFix crashes in debugger
```
Why are we adding this comment?

pratiksha-99 replied:
Removed the comment.

rniwa commented on LayoutTests/fast/repaint/repaint-renderer-with-layout-crash.html:
```diff
@@ -0,0 +1,44 @@
+<style>
```
We should simplify this test code more.

rniwa commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -1012,11 +1015,12 @@ void RenderObject::repaintUsingContainer(SingleThreadWeakPtr<const RenderLayerMo
 
     if (CheckedPtr fragmentedFlow = dynamicDowncast<RenderFragmentedFlow>(*repaintContainer)) {
-        fragmentedFlow->repaintRectangleInFragments(r);
+        if (fragmentedFlow)
```
This is redundant. line 1014/1017 already checks the nullity of `fragmentedFlow`.

pratiksha-99 replied:
got it. Removing the redundant line.

rniwa commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -1012,11 +1015,12 @@ void RenderObject::repaintUsingContainer(SingleThreadWeakPtr<const RenderLayerMo
     }
 
-    if (!repaintContainer)
+    if (!repaintContainer || r.isEmpty())
```
Is this really required to fix the crash?

pratiksha-99 replied:
This is unrelated to this crash, but might cause crashes in future so I added a check. I can remove it if you think it not needed for this radar.

rniwa replied:
I think we should revert this change unless it's needed to fix the test case.

pratiksha-99 replied:
ok sure

mdubet committed:
Added nullptr checks for RenderObject
Need the bug URL (OOPS!).
rdar://135596465.

Reviewed by NOBODY (OOPS!).

Added nullptr checks for RenderObject before de-referencing.

* LayoutTests/fast/repaint/repaint-renderer-with-layout-crash-expected.txt: Added.
* LayoutTests/fast/repaint/repaint-renderer-with-layout-crash.html: Added.
* Source/WebCore/rendering/RenderObject.cpp:
 (WebCore::RenderObject::propagateRepaintToParentWithOutlineAutoIfNeeded const):

rniwa commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -981,6 +981,9 @@ void RenderObject::propagateRepaintToParentWithOutlineAutoIfNeeded(const RenderL
             renderer = WTFMove(renderMultiColumnPlaceholder);
         }
 
+        if (!renderer || !originalRenderer)
```
Surely, `originalRenderer` can't be `nullptr` here since we just checked that it's not `nullptr` as the loop invariant?

alanbaradlay replied:
we do assert on renderMultiColumnPlaceholder a few lines above. It would be great to figure out how we end up with an invalid placeholder here and potentially fix the root cause instead.

rniwa replied:
oh yeah, we do. We should figure out why `findColumnSpannerPlaceholder` is returning `nullptr`.

pratiksha-99 replied:
~I can't get lldb to stop for the `findColumnSpannerPlaceholder()` so maybe it is not the issue.~
<img width="1061" alt="Screenshot 2024-10-06 at 7 20 48 PM" src="https://github.com/user-attachments/assets/cb3396cd-be05-49d4-ac92-b41730c5ceb9">

 There are 2 possible cases I can see:
1. `enclosingMultiColumnFlow` is nullptr - this can happen when `multiColumnFlow()` [here](https://github.com/WebKit/WebKit/blob/main/Source/WebCore/rendering/RenderBlockFlow.h#L272) returns nullptr when `hasRareBlockFlowData()` is false. This is quite possible in the fuzzing test case.
2. ~The downcast from `renderer` somehow fails.~

What should I do to handle this ? In debug build the assertion will fail in line 980 above.
